divido la tabla de clientes en dos, segun la paridad del campo ID.

dragonnegro78 opened the pull request:
incorporo chequeo de disponibilidad antes de pedir datos de los clientes.

dragonnegro78 committed:
divido la tabla de clientes en dos, segun la paridad del campo ID.
incorporo chequeo de disponibilidad antes de pedir datos de los clientes.

leamarty commented on M2/practica/clientes.php:
```diff
@@ -1,7 +1,34 @@
 <?php
 
-$json = file_get_contents('clientes.json');
+$json = file_get_contents("clientes.json");
```
Comillas simples. Everywhere please. Configurate el IDE

leamarty commented on M2/practica/clientes.php:
```diff
@@ -1,7 +1,34 @@
 
-echo json_encode($clientes);
\ No newline at end of file
+if (! (isset($_GET["filtro"]) && isset($_GET["desde"]) && isset($_GET["cantidad"]))) {
```
Ojo que esto va a salir por el .success() y va a ser una respuesta vacía, ni siquiera un json...

leamarty commented on M2/practica/clientes.php:
```diff
@@ -1,7 +1,34 @@
+}
+
+for (; $i - $_GET["desde"] < 10 && $i - 1 < count($clientes); $i += 2) {
+    $r .= json_encode($clientes[$i - 1]) . ",";
```
Entiendo que lo entiendas así, pero json_encode también puede encodear una lista, no sólo sirve para objetos. Si tenés una lista de objetos (o una lista de arrays asociativas) el json_encode también funciona. No hace falta TODO EL LIO que arranca desde la linea 11 hasta la 26. Probalo y fijate si entendés lo que digo

leamarty commented on M2/practica/index.html:
```diff
@@ -1,105 +1,183 @@
+                        $("<span>").addClass("glyphicon glyphicon-info-sign")
+                );
+
+                $.ajax({
```
Por qué no usás $.get?

dragonnegro78 replied:
@leamarty para mandar cache: false y que el navegador no recuerde la respuesta del server de autorización, pero puede que sea un error mio porque no entendí bien como habia que manejar un error. Podía haber usado post también.

leamarty replied:
Ah bueno está bien, para lo de cache está bien. Sí, podrías haber usado post, pero lo "correcto" es usar GET en estos casos... post es para crear cosas.

leamarty commented on M2/practica/index.html:
```diff
@@ -1,105 +1,183 @@
+                    type: "GET",
+                    cache: false,
+                    error: function () {
+                        $("div#div-alert").attr("class", "alert alert-danger").text(" El servidor se encuentra ocupado. Por favor, espere un momento e intente nuevamente.").prepend(
```
Mmm estás creando y borrando los divs de alert cada vez que el tipo hace click en el botón, y los textos son siempre los mismos.

Fijate que se puede hacer algo un poco más sencillo y performante, para no andar creándolos y borrándolos todo el tiempo... Te dejo el bichito de la duda: :beetle:

dragonnegro78 replied:
@leamarty no los estoy creando cada vez sino cambiando el estilo y contenido del mismo div todo el tiempo. Creo pero no estoy seguro, de que hay un flicker cuando se hace la transición. La "solución" que se me ocurre es crear todos los divs desde el html porque son todos estáticos y mantenerlos todos ocultos menos el que se quiere mostrar en ese momento. Que opinas?

leamarty replied:
Tenés razón, no lo estás creando de nuevo. Pero sí le estás cambiando el texto y la clase todo el tiempo.

Claro! Esa es la solución que decía. Dejarlos fijos y estáticos en HTML, e ir ocultando y mostrando el correspondiente.

leamarty commented on M2/practica/index.html:
```diff
@@ -1,105 +1,183 @@
+                        $("div#div-alert").attr("class", "alert alert-success").text(" El servidor esta disponible para atender el pedido.").prepend(
+                                $("<span>").addClass("glyphicon glyphicon-ok")
+                        );
+                        $.getJSON(
```
Fijate que tanto esta función como la de "par" son MUY parecidas.. todo el código desde la 44 a la 77 es casi igual que el de abajo.

Tratá de ver cómo podrías evitar repetir tanto código.

leamarty commented:
Acordate de volver a asignarme a mí cuando terminás de responder las preguntas, así sé cuándo te tengo que responder.

dragonnegro78 committed:
emprolijo clientes.php y cambio formato de respuesta para contemplar caso de que los parametros pasados sean incorrectos

dragonnegro78 committed:
clientes.php: saco una comprobacion de sanidad en clientes.php porque estaba filtrando inputs que no debian tirar error.
index.html: creo todos los alerts estaticamente y solo los muestro/oculto segun cual necesito en vez de modificar la clase y contenido de un unico alert.
reuso el codigo que carga los clientes en tablas separadas para usarlo solo 1 vez segun el id de la tabla

dragonnegro78 commented on M2/practica/index.html:
```diff
@@ -1,105 +1,171 @@
+                                                        ).children("td").addClass("vcenter").parent()
+                                                )
+                                            });
+                                            (json["clientes"][0].id % 2 == 1 ? $("table#tabla-clientes-impar") : $("table#tabla-clientes-par")).children("tbody").prepend($(tabla_temporal).html());
```
Esto use para evitar la mutabilidad

leamarty replied:
Si tenés que hacer un manejo así, entonces la función no es lo suficientemente genérica... 

Haciendo lo de arriba, lo de sacar la función, debería solucionarse esto

dragonnegro78 commented on M2/practica/index.html:
```diff
@@ -1,105 +1,171 @@
+
+                mostrar_alert("comprobando-disponibilidad");
+
+                $.ajax({
```
Quedamos en que usaba .ajax porque necesitaba cache: false

dragonnegro78 commented on M2/practica/clientes.php:
```diff
@@ -1,7 +1,32 @@
+
+for ($i = $_GET["desde"]; $i - $_GET["desde"] < $_GET["cantidad"] && $i - 1 < count($clientes); $i++) {
+    if ($i % 2 == ($_GET["filtro"] == "impar" ? 1 : 0)) {
+        $res["clientes"][] = $clientes[$i - 1];
```
Si ningun cliente satisface las condiciones, devuelve solo resultado: success. El JS interpreta que como no recibio clientes, se llego al final de la lista

leamarty commented on M2/practica/clientes.php:
```diff
@@ -1,7 +1,32 @@
\ No newline at end of file
+// falta comprobar que sean numeros
+if (!(isset($_GET["filtro"]) && isset($_GET["desde"]) && isset($_GET["cantidad"])) ||
+    ($_GET["filtro"] != "impar" && $_GET["filtro"] != "par") ||
```
`!in_array($_GET["filtro"], array('par', 'impar'))` :wink:

leamarty commented on M2/practica/clientes.php:
```diff
@@ -1,7 +1,32 @@
+);
+
+for ($i = $_GET["desde"]; $i - $_GET["desde"] < $_GET["cantidad"] && $i - 1 < count($clientes); $i++) {
+    if ($i % 2 == ($_GET["filtro"] == "impar" ? 1 : 0)) {
```
````if ($i % 2 && $_GET["filtro"]) {```` también funciona y es más cortito

leamarty commented on M2/practica/clientes.php:
```diff
@@ -1,7 +1,32 @@
+    $_GET["desde"] < 1 /* || $_GET["desde"] > count($clientes) */
+) {
+    $res = array(
+        "resultado" => "error"
```
Ojo con hacer las cosas de esta manera porque cuando haya un error va a seguir saltando por el SUCCESS en JS... podés manejarlo con los headers como dijiste en Skype

leamarty commented on M2/practica/index.html:
```diff
@@ -1,105 +1,171 @@
+                    success: function () {
+                        mostrar_alert("disponible");
+
+                        for (var paridad in ["impar", "par"]) {
```
No sirve de nada hacerlo así si después tenés que hacer esto: https://github.com/leamarty/cursoweb/pull/14/files#diff-eb05a89e75cbc7b769a2dbfc8944c2a1R51

Fijate de hacer lo que dijimos de sacar todo en una función aparte.. o con $.each en todo caso

leamarty commented:
@dragonnegro78 va queriendo... te faltan algunos retoques para que quede perfectttt

dragonnegro78 committed:
clientes.php: elimino la variable resultado que se mandaba por ajax cuando se pedian clientes y simplemente respondo como fail (401) si el resultado era error.
index.html: muevo el getjson a una funcion propia para no tener que hacer hacks por la mutabilidad de una variable.
agrego ademas caso en el que getjson devuelva error (cuando clientes.php detecta un error en los inputs)

dragonnegro78 committed:
cambio todas las comillas dobles a simples y agrego un sleep(3) antes del 401 si filtro es impar pero los otros inputs son invalidos

dragonnegro78 commented on M2/practica/index.html:
```diff
@@ -1,105 +1,172 @@
+                                        ).children("td").addClass("vcenter").parent()
+                                )
+                            });
+                            $("table#tabla-clientes-" + paridad).children("tbody").prepend($(tabla_temporal).html());
```
Necesito .html() para que no agregue el tbody, solo lo interior.
